File: backend_python/app/api/v1/hooks.py
"""API endpoints for hook management."""
from typing import Optional

from fastapi import APIRouter, HTTPException, Query

from app.models.schemas import Hook, HookCreate, HookListResponse, HookUpdate
from app.services.hook_service import HookService

router = APIRouter(prefix="/hooks", tags=["Hooks"])


@router.get("", response_model=HookListResponse)
async def list_hooks(
    project_path: Optional[str] = Query(None, description="Project path")
):
    """
    List all hooks from user and project settings files.

    Args:
        project_path: Optional project directory path

    Returns:
        List of all hooks
    """
    service = HookService()
    hooks = service.list_hooks(project_path)
    return HookListResponse(hooks=hooks)


@router.get("/{event}", response_model=HookListResponse)
async def get_hooks_by_event(
    event: str,
    project_path: Optional[str] = Query(None, description="Project path")
):
    """
    Get hooks for a specific event type.

    Args:
        event: Event type (e.g., PreToolUse, PostToolUse, Stop)
        project_path: Optional project directory path

    Returns:
        List of hooks for the specified event
    """
    service = HookService()
    hooks = service.get_hooks_by_event(event, project_path)
    return HookListResponse(hooks=hooks)


@router.post("", response_model=Hook, status_code=201)
async def create_hook(
    hook: HookCreate,
    project_path: Optional[str] = Query(None, description="Project path")
):
    """
    Create a new hook.

    Args:
        hook: Hook creation data
        project_path: Optional project directory path (required for project scope)

    Returns:
        Created hook
    """
    # Validate hook type
    if hook.type not in ["command", "prompt", "agent", "http"]:
        raise HTTPException(
            status_code=400,
            detail="Hook type must be 'command', 'prompt', 'agent', or 'http'"
        )

    # Validate scope
    if hook.scope not in ["user", "project"]:
        raise HTTPException(
            status_code=400,
            detail="Scope must be 'user' or 'project'"
        )

    # Validate required fields based on type
    if hook.type == "command" and not hook.command:
        raise HTTPException(
            status_code=400,
            detail="Command is required for command-type hooks"
        )

    if hook.type == "prompt" and not hook.prompt:
        raise HTTPException(
            status_code=400,
            detail="Prompt is required for prompt-type hooks"
        )

    if hook.type == "http":
        if not hook.url:
            raise HTTPException(
                status_code=400,
                detail="URL is required for http-type hooks"
            )
        if hook.command or hook.prompt:
            raise HTTPException(
                status_code=400,
                detail="HTTP hooks should not have command or prompt fields"
            )

    try:
        service = HookService()
        created_hook = service.add_hook(hook, project_path)
        return created_hook
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to create hook: {str(e)}"
        )


@router.put("/{hook_id}", response_model=Hook)
async def update_hook(
    hook_id: str,
    hook_update: HookUpdate,
    scope: str = Query(..., description="Scope (user or project)"),
    project_path: Optional[str] = Query(None, description="Project path")
):
    """
    Update an existing hook.

    Args:
        hook_id: ID of the hook to update
        hook_update: Hook update data
        scope: Scope of the hook (user or project)
        project_path: Optional project directory path (required for project scope)

    Returns:
        Updated hook
    """
    # Validate scope
    if scope not in ["user", "project"]:
        raise HTTPException(
            status_code=400,
            detail="Scope must be 'user' or 'project'"
        )

    # Validate hook type if provided
    if hook_update.type and hook_update.type not in ["command", "prompt", "agent", "http"]:
        raise HTTPException(
            status_code=400,
            detail="Hook type must be 'command', 'prompt', 'agent', or 'http'"
        )

    try:
        service = HookService()
        updated_hook = service.update_hook(hook_id, hook_update, scope, project_path)

        if not updated_hook:
            raise HTTPException(
                status_code=404,
                detail=f"Hook with ID {hook_id} not found in {scope} scope"
            )

        return updated_hook
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update hook: {str(e)}"
        )


@router.delete("/{hook_id}", status_code=204)
async def delete_hook(
    hook_id: str,
    scope: str = Query(..., description="Scope (user or project)"),
    project_path: Optional[str] = Query(None, description="Project path")
):
    """
    Delete a hook.

    Args:
        hook_id: ID of the hook to delete
        scope: Scope of the hook (user or project)
        project_path: Optional project directory path (required for project scope)

    Returns:
        204 No Content on success
    """
    # Validate scope
    if scope not in ["user", "project"]:
        raise HTTPException(
            status_code=400,
            detail="Scope must be 'user' or 'project'"
        )

    try:
        service = HookService()
        removed = service.remove_hook(hook_id, scope, project_path)

        if not removed:
            raise HTTPException(
                status_code=404,
                detail=f"Hook with ID {hook_id} not found in {scope} scope"
            )

        return None
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to delete hook: {str(e)}"
        )

File: app/web/src/types/hooks.ts
// Hook TypeScript types matching backend schemas

export type HookEvent =
  | "PreToolUse"
  | "PostToolUse"
  | "PostToolUseFailure"
  | "Stop"
  | "StopFailure"
  | "SessionStart"
  | "SessionEnd"
  | "UserPromptSubmit"
  | "PermissionRequest"
  | "PermissionDenied"
  | "Notification"
  | "SubagentStart"
  | "SubagentStop"
  | "TeammateIdle"
  | "TaskCreated"
  | "TaskCompleted"
  | "PreCompact"
  | "PostCompact"
  | "InstructionsLoaded"
  | "ConfigChange"
  | "CwdChanged"
  | "FileChanged"
  | "WorktreeCreate"
  | "WorktreeRemove"
  | "Elicitation"
  | "ElicitationResult";

export type HookType = "command" | "prompt" | "agent" | "http";

export interface Hook {
  id: string;
  event: HookEvent;
  matcher?: string;
  if?: string;  // Permission rule filter
  type: HookType;
  command?: string;
  prompt?: string;
  model?: string;
  async_?: boolean;  // Run in background (JSON field: "async")
  shell?: "bash" | "powershell";
  statusMessage?: string;
  once?: boolean;
  timeout?: number;
  url?: string;
  headers?: Record<string, string>;
  allowedEnvVars?: string[];
  scope: "user" | "project";
}

export interface HookCreate {
  event: HookEvent;
  matcher?: string;
  type: HookType;
  command?: string;
  prompt?: string;
  model?: string;
  async_?: boolean;
  statusMessage?: string;
  once?: boolean;
  timeout?: number;
  url?: string;
  headers?: Record<string, string>;
  allowedEnvVars?: string[];
  scope: "user" | "project";
}

export interface HookUpdate {
  event?: HookEvent;
  matcher?: string;
  type?: HookType;
  command?: string;
  prompt?: string;
  model?: string;
  async_?: boolean;
  statusMessage?: string;
  once?: boolean;
  timeout?: number;
  url?: string;
  headers?: Record<string, string>;
  allowedEnvVars?: string[];
}

export interface HookListResponse {
  hooks: Hook[];
}

// UI-specific types
export interface HookCardProps {
  hook: Hook;
  onEdit: (hook: Hook) => void;
  onDelete: (hookId: string, scope: "user" | "project") => void;
}

export interface HookListProps {
  hooks: Hook[];
  onEdit: (hook: Hook) => void;
  onDelete: (hookId: string, scope: "user" | "project") => void;
}

// Event type metadata for UI
export interface HookEventMetadata {
  name: HookEvent;
  label: string;
  description: string;
  icon: string;
}

export const HOOK_EVENTS: HookEventMetadata[] = [
  {
    name: "PreToolUse",
    label: "Pre-Tool Use",
    description: "Triggered before Claude executes a tool",
    icon: "🔧",
  },
  {
    name: "PostToolUse",
    label: "Post-Tool Use",
    description: "Triggered after Claude executes a tool",
    icon: "✅",
  },
  {
    name: "PostToolUseFailure",
    label: "Post-Tool Use Failure",
    description: "Triggered when a tool execution fails",
    icon: "❌",
  },
  {
    name: "Stop",
    label: "Stop",
    description: "Triggered after Claude responds",
    icon: "🛑",
  },
  {
    name: "StopFailure",
    label: "Stop Failure",
    description: "Triggered on API error during a turn",
    icon: "💥",
  },
  {
    name: "SessionStart",
    label: "Session Start",
    description: "Triggered when a session starts",
    icon: "🚀",
  },
  {
    name: "SessionEnd",
    label: "Session End",
    description: "Triggered when a session ends",
    icon: "🏁",
  },
  {
    name: "UserPromptSubmit",
    label: "User Prompt Submit",
    description: "Triggered when user submits a prompt",
    icon: "💬",
  },
  {
    name: "PermissionRequest",
    label: "Permission Request",
    description: "Triggered when a permission dialog appears",
    icon: "🔐",
  },
  {
    name: "PermissionDenied",
    label: "Permission Denied",
    description: "Triggered when auto mode denies a tool",
    icon: "🚫",
  },
  {
    name: "Notification",
    label: "Notification",
    description: "Triggered on notifications",
    icon: "🔔",
  },
  {
    name: "SubagentStart",
    label: "Subagent Start",
    description: "Triggered when a subagent starts",
    icon: "🤖",
  },
  {
    name: "SubagentStop",
    label: "Subagent Stop",
    description: "Triggered when a subagent finishes",
    icon: "🤖",
  },
  {
    name: "TeammateIdle",
    label: "Teammate Idle",
    description: "Triggered when a team member becomes idle",
    icon: "💤",
  },
  {
    name: "TaskCreated",
    label: "Task Created",
    description: "Triggered when a task is created via TaskCreate",
    icon: "📋",
  },
  {
    name: "TaskCompleted",
    label: "Task Completed",
    description: "Triggered when a task is marked complete",
    icon: "✔️",
  },
  {
    name: "PreCompact",
    label: "Pre-Compact",
    description: "Triggered before context compaction",
    icon: "📦",
  },
  {
    name: "PostCompact",
    label: "Post-Compact",
    description: "Triggered after context compaction",
    icon: "📦",
  },
  {
    name: "InstructionsLoaded",
    label: "Instructions Loaded",
    description: "Triggered when CLAUDE.md or rules are loaded",
    icon: "📄",
  },
  {
    name: "ConfigChange",
    label: "Config Change",
    description: "Triggered when a settings file changes",
    icon: "⚙️",
  },
  {
    name: "CwdChanged",
    label: "CWD Changed",
    description: "Triggered when the working directory changes",
    icon: "📂",
  },
  {
    name: "FileChanged",
    label: "File Changed",
    description: "Triggered when a watched file changes",
    icon: "📝",
  },
  {
    name: "WorktreeCreate",
    label: "Worktree Create",
    description: "Triggered when a git worktree is created",
    icon: "🌳",
  },
  {
    name: "WorktreeRemove",
    label: "Worktree Remove",
    description: "Triggered when a git worktree is removed",
    icon: "🌳",
  },
  {
    name: "Elicitation",
    label: "Elicitation",
    description: "Triggered on MCP elicitation input",
    icon: "❓",
  },
  {
    name: "ElicitationResult",
    label: "Elicitation Result",
    description: "Triggered when MCP elicitation result arrives",
    icon: "❓",
  },
];

// Model options for agent hooks
export const AGENT_MODELS = [
  { value: "haiku", label: "Haiku (Fast)" },
  { value: "sonnet", label: "Sonnet (Balanced)" },
  { value: "opus", label: "Opus (Powerful)" },
];

// Matcher pattern examples
export const MATCHER_EXAMPLES = [
  { pattern: "Write(*.py)", description: "Matches Write tool on Python files" },
  { pattern: "Bash", description: "Matches any Bash tool use" },
  { pattern: "Edit|Write", description: "Matches Edit OR Write tool" },
  { pattern: "Read(*.md)", description: "Matches Read tool on Markdown files" },
  { pattern: "*", description: "Matches all tools" },
];

// Environment variables available in hooks
export const HOOK_ENV_VARS = [
  {
    name: "$CLAUDE_FILE_PATHS",
    description: "File paths involved in the tool use",
  },
  { name: "$CLAUDE_TOOL_NAME", description: "Name of the tool being executed" },
  {
    name: "$CLAUDE_TOOL_ARGS",
    description: "Arguments passed to the tool (JSON)",
  },
  { name: "$CLAUDE_PROJECT_PATH", description: "Current project directory path" },
  { name: "$CLAUDE_USER_HOME", description: "User's home directory" },
];

// Hook templates for quick creation
export interface HookTemplate {
  name: string;
  description: string;
  event: HookEvent;
  type: HookType;
  matcher?: string;
  command?: string;
  prompt?: string;
  model?: string;
  async_?: boolean;
  statusMessage?: string;
  once?: boolean;
  timeout?: number;
  url?: string;
}

export const HOOK_TEMPLATES: HookTemplate[] = [
  {
    name: "Blank Hook",
    description: "Start from scratch",
    event: "PreToolUse",
    type: "command",
    command: "",
  },
  {
    name: "Python File Linter",
    description: "Run linter before writing Python files",
    event: "PreToolUse",
    type: "command",
    matcher: "Write(*.py)",
    command: "python -m pylint $CLAUDE_FILE_PATHS",
    timeout: 10,
  },
  {
    name: "Git Auto-commit",
    description: "Commit changes after file writes",
    event: "PostToolUse",
    type: "command",
    matcher: "Write|Edit",
    command: 'git add $CLAUDE_FILE_PATHS && git commit -m "Auto-commit: $CLAUDE_TOOL_NAME"',
    timeout: 5,
  },
  {
    name: "Security Reminder",
    description: "Remind about security when reading config files",
    event: "PreToolUse",
    type: "prompt",
    matcher: "Read(*.env|*.json)",
    prompt:
      "Remember to mask any sensitive data like API keys, tokens, or passwords when displaying file contents to the user.",
  },
  {
    name: "Session Start Logger",
    description: "Log session start time",
    event: "SessionStart",
    type: "command",
    command: 'echo "Session started at $(date)" >> ~/.claude/session.log',
    timeout: 2,
  },
  {
    name: "Code Review Agent",
    description: "Use an agent to review code changes",
    event: "PostToolUse",
    type: "agent",
    matcher: "Write(*.ts|*.tsx|*.js|*.jsx)",
    prompt: "Review this code change for best practices, potential bugs, and security issues. Provide brief feedback.",
    model: "haiku",
    async_: true,
    statusMessage: "Reviewing code...",
  },
  {
    name: "Session Summary Agent",
    description: "Generate session summary on end",
    event: "SessionEnd",
    type: "agent",
    prompt: "Summarize the key accomplishments and changes made during this session.",
    model: "haiku",
    once: true,
  },
  {
    name: "Presence Dashboard Logger",
    description: "Send events to Presence Dashboard via HTTP hook",
    event: "PostToolUse",
    type: "http",
    url: "http://localhost:8000/api/v1/presence/events",
  },
];
